json-decode.go

```diff
@@ -12,65 +12,52 @@ import (
 //----------------------------------------
 // cdc.decodeReflectJSON
 
-// CONTRACT: rv is addressable as per https://golang.org/pkg/reflect/#Value.CanAddr
-func (cdc *Codec) decodeReflectJSON(bz []byte, info *TypeInfo, rv reflect.Value, opts FieldOptions) error {
-	// By this point, we MUST be dealing with an addressable value.
+// CONTRACT: rv.CanAddr() is true.
+func (cdc *Codec) decodeReflectJSON(bz []byte, info *TypeInfo, rv reflect.Value, opts FieldOptions) (err error) {
 	if !rv.CanAddr() {
-		panic("value is not addressable")
+		panic("rv not addressable")
 	}
-
-	// SANITY CHECK
 	if info.Type.Kind() == reflect.Interface && rv.Kind() == reflect.Ptr {
 		panic("should not happen")
 	}
 
-	// No need for diambiguation, decode as is.
+	// No need for disambiguation, decode as is.
 	if !info.Registered {
-		return cdc._decodeReflectJSON(bz, info, rv, opts)
+		err = cdc._decodeReflectJSON(bz, info, rv, opts)
+		return
 	}
 
-	// Otherwise disambiguation time
-	disfix, restBlob, err := parseDisfixAndData(bz)
+	// Otherwise, disambiguation time.
+	disfix, bz, err := decodeDisfixJSON(bz)
 	if err != nil {
-		return err
+		return
 	}
-	info, err = cdc.getTypeInfoFromDisfix_rlock(disfix)
+	cinfo, err := cdc.getTypeInfoFromDisfix_rlock(disfix)
 	if err != nil {
-		return err
+		return
 	}
 
-	bz = restBlob
 	// And we need to construct the concrete type
 	// that'll then be set into the interface field.
-	crv, _ := constructConcreteType(info)
-	if err := cdc._decodeReflectJSON(bz, info, crv, opts); err != nil {
-		return err
+	var crv, _ = constructConcreteType(cinfo)
+	if err = cdc._decodeReflectJSON(bz, info, crv, opts); err != nil {
+		rv.Set(crv) // Helps with debugging
+		return
 	}
-	rv.Set(crv)
-	return nil
-}
 
-// constructConcreteType creates the concrete value as
-// well as the corresponding settable value for it.
-func constructConcreteType(cinfo *TypeInfo) (crv, rvSet reflect.Value) {
-	// Construct new concrete type.
-	// NOTE: rv.Set() should succeed because it was validated
-	// already during Register[Interface/Concrete].
-	if cinfo.PointerPreferred {
-		cPtrRv := reflect.New(cinfo.Type)
-		crv = cPtrRv.Elem()
-	} else {
-		crv = reflect.New(cinfo.Type).Elem()
-		rvSet = crv
-	}
-	return crv, rvSet
+	// We need to set here, for when !PointerPreferred and the type
+	// is say, an array of bytes (e.g. [32]byte), then we must call
+	// rv.Set() *after* the value was acquired.
+	rv.Set(crv)
+	return
 }
 
 func (cdc *Codec) _decodeReflectJSON(bz []byte, info *TypeInfo, rv reflect.Value, opts FieldOptions) error {
+
 	// If the type implements json.Unmarshaler, just
 	// automatically respect that and skip to it.
-	if ok, err := processIfUnmarshaler(bz, rv); ok || err != nil {
```

Comment:
I feel like the indirection through a function is unnecessary...
"if reflect-value's addr (pointer) type implements unmarshaller type" reads itself ok...

Reply:
That checked if both the basic pointer and it's pointer-pointer implemented Unmarshal, for example if you had this code https://play.golang.org/p/tX3XlfoEtmX
```go
package main

import (
	"encoding/json"
	"fmt"
	"log"
)

type ts string

func (t *ts) UnmarshalJSON(b []byte) error {
	*t = ts(b)
	return nil
}

func main() {
	t1 := ts("")
	t2 := &t1
	if err := json.Unmarshal([]byte(`"foo"`), &t2); err != nil { // Notice the pointer-pointer passed in?
		log.Fatal(err)
	}
	fmt.Printf("%s\n", t1)
}
```
we MUST be able to get back "foo"

Reply:
Outside of extracting that code into a helper function, we'd clutter our main code with all the that specialized code.

Reply:
I don't see the clutter, I see less clutter here, it's a single line either way, and spelling it out is more descriptive.

If it turned into 2 lines, say, then I would prefer the helper function.  Yak shaving, as Ethan says.

Reply:
The code for processIfUnmarshaler isn't 1 or 2 lines though, it is about 17 lines, that's what I meant by clutter.

Reply:
Well, I see your point too.